### PR TITLE
feat: added connect wallet feature to Dapp

### PR DESCRIPTION
**Pull request by quiet-node**
**Description**: This PR adds the connect wallet feature to the dapp

UI flow:

a. If the current wallet does not match the expected Hedera networks, a toast will pop up to prompt users to follow the appropriate guidelines for user and network configuration.
image

b. When the network is configured properly, request accounts by calling eth_requestAccounts

image

image

c. After connecting the account, the information will be stored in Cookies. At the same time, the user will be navigated to the first protected route "/overview".

image

Fixes: #277

### CODE REVIEW

File: system-contract-dapp-playground/src/components/toast/CommonToast.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    ),
+  });
+};
+
+export const NoEthToast = ({ toaster }: PageProps) => {
+  toaster({
+    position: 'top',
+    render: () => (
+      <div className="text-white text-center font-styrene bg-black px-6 py-3 rounded-xl tracking-wide">
+        <div className="font-medium">🚨 No Ethereum wallet detected 🚨</div>
```
Comment (Nana-EC) [inline system-contract-dapp-playground/src/components/toast/CommonToast.tsx:50]:
```suggestion
        <div className="font-medium">🚨 No Hedera wallet detected 🚨</div>
```

File: system-contract-dapp-playground/src/components/toast/CommonToast.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    render: () => (
+      <div className="text-white text-center font-styrene bg-black px-6 py-3 rounded-xl tracking-wide flex flex-col gap-1">
+        <div className="font-medium">🚨 {title} 🚨</div>
+        <div>{description}</div>
+      </div>
+    ),
+  });
+};
+
+export const NoEthToast = ({ toaster }: PageProps) => {
```
Comment (Nana-EC) [inline system-contract-dapp-playground/src/components/toast/CommonToast.tsx:45]:
```suggestion
export const NoHbarToast = ({ toaster }: PageProps) => {
```

File: system-contract-dapp-playground/src/sections/landing/index.tsx [hunk 3]
```diff
@@ -21,10 +21,79 @@
...
+  const router = useRouter();
+  const toaster = useToast();
+  const [accounts, setAccounts] = useState<string[]>([]);
   const verticalVariant = VerticalCommonVariants(30, 0.5);
+  const { walletProvider, err: walletProviderErr } = getWalletProvider();
+
+  /** @dev handle connect wallet when a user click `connect wallet` button */
+  const handleConnectWallet = async () => {
+    // handle ethereum object or walletProvider being null by toasting it out on the client
+    if (walletProviderErr === '!ETHEREUM' || !walletProvider) {
```
Comment (Nana-EC) [inline system-contract-dapp-playground/src/sections/landing/index.tsx:42]: Is this string hardcoded to Metamask?
Should it be Hedera?

Either ways pull it out to a const of a class variable

File: system-contract-dapp-playground/src/sections/landing/index.tsx [hunk 4]
```diff
@@ -21,10 +21,79 @@
...
+    }
+
+    // call requestAccount() API for users to connect their account to the DApp
+    const { accounts, err: getAccountErr } = await requestAccount(walletProvider!);
+
+    // handle getAccountError
+    if (getAccountErr || !accounts || accounts.length === 0) {
+      let errorMessage = 'Unknown error appeared...';
+
+      if (JSON.stringify(getAccountErr).indexOf('4001') !== -1) {
```
Comment (Nana-EC) [inline system-contract-dapp-playground/src/sections/landing/index.tsx:60]: Q: why index 4001? Are you searching for the error code?
If so add a comment

File: system-contract-dapp-playground/src/sections/landing/index.tsx [hunk 5]
```diff
@@ -21,10 +21,79 @@
...
+        errorMessage = 'You have rejected the request.';
+      } else if (JSON.stringify(getAccountErr).indexOf('-32002') !== -1) {
```
Comment (Nana-EC) [inline system-contract-dapp-playground/src/sections/landing/index.tsx:62]: Q: are you searching for the error code? If so add a comment

File: system-contract-dapp-playground/src/middleware.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+import { NextResponse } from 'next/server';
+import type { NextRequest } from 'next/server';
+
+const isProtectedRoute = (pathname: string) => {
+  const protectedRoutes = [
+    '/overview',
+    '/hts-hip-206',
+    '/hrc-hip-719',
```
Comment (Nana-EC) [inline system-contract-dapp-playground/src/middleware.ts:28]:
```suggestion
    '/hrc-719',
```

File: system-contract-dapp-playground/src/middleware.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    return NextResponse.redirect(new URL(`/`, request.url));
+  }
+}
+
+export const config = {
+  matcher: [
+    '/',
+    '/overview',
+    '/hts-hip-206',
+    '/hrc-hip-719',
```
Comment (Nana-EC) [inline system-contract-dapp-playground/src/middleware.ts:56]:
```suggestion
    '/hrc-719',
```

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good.
Some questions and feedback

File: system-contract-dapp-playground/src/api/wallet/index.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+ *
+ * @params walletProvider: ethers.BrowserProvider
+ *
+ * @returns Promise<WalletResult>
+ */
+export const requestAccount = async (
+  walletProvider: ethers.BrowserProvider
+): Promise<WalletResult> => {
+  try {
+    const accounts: [string] = await walletProvider.send('eth_requestAccounts', []);
```
[the remark this answers is not in the document]
Comment (quiet-node) [reply system-contract-dapp-playground/src/api/wallet/index.ts:104]: yeah `eth_requestAccounts` is a special JSON RPC from metamask. 
https://docs.metamask.io/wallet/reference/rpc-api/#eth_requestaccounts

File: system-contract-dapp-playground/src/sections/landing/index.tsx (hunk 3 above)
Comment (quiet-node) [reply system-contract-dapp-playground/src/sections/landing/index.tsx:42]: Oh this is customized by me. You can find it here https://github.com/hashgraph/hedera-smart-contracts/blob/dapp-connect-wallet-feat/system-contract-dapp-playground/src/api/wallet/index.ts#L46

I'll switch it to Hedera!

File: system-contract-dapp-playground/src/sections/landing/index.tsx (hunk 4 above)
Comment (quiet-node) [reply system-contract-dapp-playground/src/sections/landing/index.tsx:60]: Oh yeah these are special indexes from Metamask. https://docs.metamask.io/wallet/reference/provider-api/#errors 
I'll add a comment to these! I really thought I did tho...

**Review by Nana-EC — APPROVED** [review]
LG

**Pull request merged**